### DISCUSSION TITLE
Feature Request: Levenshtein distance with bounds

### DISCUSSION THREAD

**Discussion by zmbc** [duckdb/duckdb#10345] [Ideas]
**Why do you want this feature?**

String distances such as the Levenshtein distance are fairly computationally expensive. For many real-world applications, it isn't necessary to compute them exactly. We may care what the precise Levenshtein distances are for very similar strings but any string with distance > 5 is different enough that we don't care -- it's just completely different.

My understanding is that Levenshtein distance can be significantly faster (especially for long strings, and when the "don't care" cutoff is low) if you use the cutoff in the Levenshtein algorithm itself. See for example [this StackOverflow answer](https://stackoverflow.com/a/48902758/).

My thinking is that the Levenshtein function could have an optional argument that specifies the cutoff, with the default remaining no cutoff.

Is this something the DuckDB team would be interested in? I could probably contribute a PR.

**Comment by zmbc** [comment]
Bumping this -- **Is this something the DuckDB team would be interested in?**

**Comment by soerenwolfers** [comment]
In the meantime, can you just do `CASE IF |len(a) - len(b)| > 5 THEN 'Infinity' ELSE levenshtein(a, b) END`?

That's so trivial that it's questionable whether it's worth a new function. If, on the other hand, there was an approximate algorithm to compute Levenstein distances with a bounded relative error in a reduced computational complexity class, that would be a much more interesting thing to provide, in my opinion.

**Reply by zmbc** [reply]
No, the optimization available is much more than the length check you suggest. For example "abcdefghijkl" and "xxxdfghijxkl" have identical lengths and Levenshtein distance 5, but you need only look at the first 3 characters to tell that the Levenshtein distance has to be at _least_ 3.

I'm surprised not to be able to immediately find a formal write-up of what seems to me like a very intuitive optimization; perhaps the bounded Levenshtein is not very relevant outside of a few applications. The one I have in mind is record linkage, where it is _extremely_ relevant.

Would it help for me to write a Python and/or pseudocode implementation to formalize this?

**Reply by soerenwolfers** [reply]
Ah, sorry, the linked stack overflow answer didn't make that as clear. I think adding static element wise functions is probably the easiest thing you could change about duckdb. Why not make a PR following along the design of the original distance at https://github.com/duckdb/duckdb/blob/c5e173a5bcd0839a7ecccad785bcae7f50c6fde3/src/core_functions/scalar/string/levenshtein.cpp#L15

I'm sure if you have remaining questions (like how to wire you the extra integer argument) the maintainers would be happy to help get your PR over the line.

**Reply by zmbc** [reply]
@soerenwolfers Do you know which maintainer to ping about https://github.com/duckdb/duckdb/pull/12079? It's been sitting open for quite some time without review.

**Reply by soerenwolfers** [reply]
@szarnyasg might have an opinion. it might also be true that the devs are waiting for the CI/CD not to fail before reviewing (you have some unit tests failing, see https://github.com/duckdb/duckdb/actions/runs/10257134636/job/28377592054?pr=12079) , or possibly there is a mailing list or discord channel where you can ask for a review.

**Reply by zmbc** [reply]
@szarnyasg The CI failures on the PR are now resolved, would you be able to take a look? #12079

**Comment by cmdlineluser** [comment]
In Python I had been using RapidFuzz - which is a wrapper around its C++ implementation.

https://github.com/rapidfuzz/rapidfuzz-cpp/blob/main/rapidfuzz/distance/Levenshtein.hpp

I have little knowledge on the subject so I may be wrong, but I think this performs the optimization you're talking about?

Perhaps it could serve as inspiration (or maybe a duckdb rapidfuzz extension could be feasible?)

**Reply by zmbc** [reply]
Yes, I had forgotten about rapidfuzz -- it does this optimization and a bunch more. It would be really interesting to use this in DuckDB!

**Reply by zmbc** [reply]
On a bit more poking, DuckDB already uses (an older version of) rapidfuzz for the Jaro and Jaro-Winkler distance functions, and these implementations _already_ support a score cutoff with accompanying optimization. The only thing to do for those is allow the user to pass it. I'll start there.
